File: electron/services/automation-executor.ts
import { BrowserWindow } from "electron";
import { spawnClaude } from "./claude-cli";
import { automationScheduler } from "./automation-scheduler";
import { storage } from "./storage";
import { skillsEngine } from "./skills-engine";
import { worktreeManager } from "./worktree-manager";
import { logger } from "../utils/logger";
import type {
  Automation,
  AutomationRun,
  AppNotification,
} from "../../src/types/ipc";

type SandboxPolicy = "read-only" | "workspace-write" | "full";

interface ExecutionContext {
  runId: string;
  automation: Automation;
  worktreePath: string | null;
  sandboxPolicy: SandboxPolicy;
  kill: (() => void) | null;
}

/**
 * Automation Executor handles running automation triggers as agent tasks.
 * Connects the scheduler to the agent orchestrator, manages worktrees,
 * captures results, and sends notifications.
 */
class AutomationExecutor {
  private activeRuns: Map<string, ExecutionContext> = new Map();

  /**
   * Initialize the executor by hooking into the scheduler's event system.
   */
  init(): void {
    automationScheduler.onEvent((type, automation, run) => {
      if (type === "triggered" && run) {
        this.executeRun(automation, run).catch((err) => {
          logger.error(
            "automation-executor",
            `Execution failed for ${automation.name}`,
            err
          );
        });
      }
    });

    logger.info("automation-executor", "Initialized");
  }

  /**
   * Execute an automation run.
   */
  async executeRun(
    automation: Automation,
    run: AutomationRun
  ): Promise<void> {
    const sandboxPolicy = (automation.triggerConfig.sandboxPolicy as SandboxPolicy) ?? "workspace-write";

    let worktreePath: string | null = null;

    try {
      // 1. Create a fresh git worktree (for git projects)
      const project = storage.getProject(automation.projectId);
      if (project) {
        try {
          const wt = await worktreeManager.create(
            project.path,
            `auto-${run.id.slice(0, 8)}`,
            `Automation: ${automation.name}`
          );
          worktreePath = wt.path;
          logger.info(
            "automation-executor",
            `Created worktree for run ${run.id}: ${wt.path}`
          );
        } catch (err) {
          logger.warn(
            "automation-executor",
            "Failed to create worktree, using project directory",
            err
          );
        }
      }

      const ctx: ExecutionContext = {
        runId: run.id,
        automation,
        worktreePath,
        sandboxPolicy,
        kill: null,
      };
      this.activeRuns.set(run.id, ctx);

      // 2. Load associated skills and build system prompt
      let systemPrompt = "";
      const allowedTools = this.getToolsForPolicy(sandboxPolicy);

      if (automation.skillIds.length > 0) {
        for (const skillId of automation.skillIds) {
          const skill = skillsEngine.resolve(skillId);
          if (skill) {
            const config = skillsEngine.invoke(skill);
            systemPrompt += config.systemPrompt + "\n\n";
            // Merge tools from skills (filtered by sandbox policy)
            for (const tool of config.tools) {
              if (
                !allowedTools.includes(tool) &&
                this.isToolAllowed(tool, sandboxPolicy)
              ) {
                allowedTools.push(tool);
              }
            }
          }
        }
      }

      // 3. Build the prompt
      const fullPrompt = systemPrompt
        ? `${systemPrompt}\n---\n\n${automation.prompt}`
        : automation.prompt;

      // 4. Launch agent via CLI spawn
      const cwd = worktreePath ?? project?.path ?? process.cwd();

      let claudeCliPath: string | undefined;
      try {
        const settings = storage.getAppSettings();
        claudeCliPath = settings.claudeCliPath ?? undefined;
      } catch {
        // ignore
      }

      const { events, kill } = spawnClaude({
        prompt: fullPrompt,
        cwd,
        permissionMode:
          sandboxPolicy === "read-only" ? "plan" : "acceptEdits",
        allowedTools,
        claudePath: claudeCliPath,
      });

      ctx.kill = kill;

      const conversation: Array<{ role: string; content: string }> = [];
      let totalCostUsd = 0;

      for await (const message of events) {
        switch (message.type) {
          case "assistant": {
            const content = message.message?.content;
            if (Array.isArray(content)) {
              const text = content
                .filter(
                  (b: Record<string, unknown>) => b.type === "text"
                )
                .map((b: Record<string, unknown>) => b.text as string)
                .join("");
              if (text) {
                conversation.push({ role: "assistant", content: text });
              }
            }
            break;
          }
          case "result": {
            if (message.result) {
              conversation.push({ role: "assistant", content: message.result });
            }
            totalCostUsd = message.total_cost_usd ?? 0;
            break;
          }
        }
      }

      // 5. Store results in automation_runs table
      const result: Record<string, unknown> = {
        conversation,
        totalCostUsd,
        worktreePath,
        sandboxPolicy,
      };

      // Determine if there's actionable output
      const hasActionableOutput = conversation.length > 0 &&
        conversation.some((c) => c.content.length > 50);

      const finalStatus = hasActionableOutput ? "completed" : "completed";
      automationScheduler.completeRun(run.id, finalStatus, result);

      // 6. Notify user
      this.sendNotification(
        automation,
        run.id,
        "success",
        `Automation "${automation.name}" completed`
      );

      // 7. Auto-archive if no actionable output
      if (!hasActionableOutput) {
        storage.updateAutomationRun(run.id, { status: "archived" });
      }

      logger.info(
        "automation-executor",
        `Run ${run.id} completed for ${automation.name}`
      );
    } catch (err) {
      const message =
        err instanceof Error ? err.message : "Unknown execution error";

      automationScheduler.completeRun(run.id, "failed", {
        error: message,
      });

      this.sendNotification(
        automation,
        run.id,
        "error",
        `Automation "${automation.name}" failed: ${message}`
      );

      logger.error(
        "automation-executor",
        `Run ${run.id} failed for ${automation.name}`,
        err
      );
    } finally {
      this.activeRuns.delete(run.id);
    }
  }

  /**
   * Get the list of running automation runs.
   */
  getActiveRuns(): ExecutionContext[] {
    return Array.from(this.activeRuns.values());
  }

  /**
   * Get the tools allowed based on the sandbox policy.
   */
  private getToolsForPolicy(policy: SandboxPolicy): string[] {
    switch (policy) {
      case "read-only":
        return ["Read", "Glob", "Grep"];
      case "workspace-write":
        return ["Read", "Edit", "Write", "Glob", "Grep", "Bash"];
      case "full":
        return [
          "Read",
          "Edit",
          "Write",
          "Bash",
          "Glob",
          "Grep",
          "WebSearch",
          "WebFetch",
        ];
    }
  }

  /**
   * Check if a tool is allowed under the given sandbox policy.
   */
  private isToolAllowed(tool: string, policy: SandboxPolicy): boolean {
    const allowed = this.getToolsForPolicy(policy);
    return allowed.includes(tool);
  }

  /**
   * Send a notification to all renderer windows.
   */
  private sendNotification(
    automation: Automation,
    runId: string,
    type: "success" | "error",
    message: string
  ): void {
    const notification: AppNotification = {
      id: crypto.randomUUID(),
      type: type === "success" ? "success" : "error",
      title: automation.name,
      message,
      timestamp: new Date().toISOString(),
    };

    const windows = BrowserWindow.getAllWindows();
    for (const win of windows) {
      win.webContents.send("notification", notification);
    }
  }
}

export const automationExecutor = new AutomationExecutor();

File: src-tauri/src/services/automation_executor.rs
use tauri::{AppHandle, Emitter};
use tracing::{error, info, warn};

use crate::db::models::{AppNotification, Automation, AutomationRun};
use crate::db::queries::{automations as auto_queries, projects as project_queries, settings as settings_queries};
use crate::error::{AppError, AppResult};
use crate::state::AppState;

use super::automation_scheduler;
use super::claude_cli::{spawn_claude, SpawnClaudeOptions};
use super::skills_engine;
use super::worktree_manager;

/// Sandbox policy determines what tools the automation agent may use.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum SandboxPolicy {
    ReadOnly,
    WorkspaceWrite,
    Full,
}

impl SandboxPolicy {
    fn from_str(s: &str) -> Self {
        match s {
            "read-only" => Self::ReadOnly,
            "full" => Self::Full,
            _ => Self::WorkspaceWrite,
        }
    }

    fn tools(&self) -> Vec<String> {
        match self {
            Self::ReadOnly => vec![
                "Read".into(),
                "Glob".into(),
                "Grep".into(),
            ],
            Self::WorkspaceWrite => vec![
                "Read".into(),
                "Edit".into(),
                "Write".into(),
                "Glob".into(),
                "Grep".into(),
                "Bash".into(),
            ],
            Self::Full => vec![
                "Read".into(),
                "Edit".into(),
                "Write".into(),
                "Bash".into(),
                "Glob".into(),
                "Grep".into(),
                "WebSearch".into(),
                "WebFetch".into(),
            ],
        }
    }

    fn is_tool_allowed(&self, tool: &str) -> bool {
        self.tools().iter().any(|t| t == tool)
    }
}

/// Execute an automation run.
///
/// This spawns a Claude CLI process, captures its conversational output,
/// stores results in the automation_runs table, and sends notifications.
pub async fn execute_run(
    app_handle: AppHandle,
    state: &AppState,
    automation: &Automation,
    run: &AutomationRun,
) -> AppResult<()> {
    let sandbox_policy_str = automation
        .trigger_config
        .get("sandboxPolicy")
        .and_then(|v| v.as_str())
        .unwrap_or("workspace-write");
    let sandbox_policy = SandboxPolicy::from_str(sandbox_policy_str);

    let mut worktree_path: Option<String> = None;

    let result: Result<(), AppError> = async {
        // 1. Create a fresh git worktree (for git projects)
        let project = {
            let db = state.db.lock();
            project_queries::get_project(&db, &automation.project_id)
                .map_err(AppError::from)?
        };

        if let Some(ref project) = project {
            match worktree_manager::create(
                &project.path,
                &format!("auto-{}", &run.id[..8.min(run.id.len())]),
                Some(&format!("Automation: {}", automation.name)),
            )
            .await
            {
                Ok(wt) => {
                    worktree_path = Some(wt.path.clone());
                    info!(
                        target: "automation-executor",
                        "Created worktree for run {}: {}", run.id, wt.path
                    );
                }
                Err(e) => {
                    warn!(
                        target: "automation-executor",
                        "Failed to create worktree, using project directory: {}", e
                    );
                }
            }
        }

        // 2. Load associated skills and build system prompt
        let mut system_prompt = String::new();
        let mut allowed_tools = sandbox_policy.tools();

        if !automation.skill_ids.is_empty() {
            let db = state.db.lock();
            for skill_id in &automation.skill_ids {
                if let Ok(Some(skill)) = skills_engine::resolve(&db, skill_id) {
                    if let Ok(config) = skills_engine::invoke(&skill) {
                        system_prompt.push_str(&config.system_prompt);
                        system_prompt.push_str("\n\n");

                        for tool in &config.tools {
                            if !allowed_tools.contains(tool)
                                && sandbox_policy.is_tool_allowed(tool)
                            {
                                allowed_tools.push(tool.clone());
                            }
                        }
                    }
                }
            }
        }

        // 3. Build the prompt
        let full_prompt = if !system_prompt.is_empty() {
            format!("{}\n---\n\n{}", system_prompt, automation.prompt)
        } else {
            automation.prompt.clone()
        };

        // 4. Launch agent via CLI spawn
        let cwd = worktree_path
            .as_deref()
            .or(project.as_ref().map(|p| p.path.as_str()))
            .unwrap_or(".")
            .to_string();

        let claude_cli_path = {
            let db = state.db.lock();
            settings_queries::get_app_settings(&db)
                .ok()
                .and_then(|s| s.claude_cli_path)
        };

        let permission_mode = match sandbox_policy {
            SandboxPolicy::ReadOnly => "plan",
            _ => "acceptEdits",
        };

        let spawn_result = spawn_claude(SpawnClaudeOptions {
            prompt: full_prompt,
            cwd,
            model: None,
            resume_session_id: None,
            permission_mode: Some(permission_mode.to_string()),
            allowed_tools: Some(allowed_tools),
            disallowed_tools: None,
            append_system_prompt: None,
            claude_path: claude_cli_path,
        })
        .map_err(|e| AppError::Agent(format!("Failed to spawn Claude CLI: {}", e)))?;

        let mut events_rx = spawn_result.events_rx;
        // We don't need stdin_tx for automations, and kill_tx is for cancellation
        let _kill_tx = spawn_result.kill_tx;

        let mut conversation: Vec<serde_json::Value> = Vec::new();
        let mut total_cost_usd: f64 = 0.0;

        while let Some(event) = events_rx.recv().await {
            match event.event_type.as_str() {
                "assistant" => {
                    if let Some(ref msg) = event.message {
                        let text: String = msg
                            .content
                            .iter()
                            .filter(|b| b.block_type == "text")
                            .filter_map(|b| b.text.as_deref())
                            .collect::<Vec<_>>()
                            .join("");
                        if !text.is_empty() {
                            conversation.push(serde_json::json!({
                                "role": "assistant",
                                "content": text,
                            }));
                        }
                    }
                }
                "result" => {
                    if let Some(ref result_text) = event.result {
                        conversation.push(serde_json::json!({
                            "role": "assistant",
                            "content": result_text,
                        }));
                    }
                    total_cost_usd = event.total_cost_usd.unwrap_or(0.0);
                }
                _ => {}
            }
        }

        // 5. Store results in automation_runs table
        let result_data = serde_json::json!({
            "conversation": conversation,
            "totalCostUsd": total_cost_usd,
            "worktreePath": worktree_path,
            "sandboxPolicy": sandbox_policy_str,
        });

        let has_actionable_output =
            !conversation.is_empty()
                && conversation.iter().any(|c| {
                    c.get("content")
                        .and_then(|v| v.as_str())
                        .map(|s| s.len() > 50)
                        .unwrap_or(false)
                });

        {
            let db = state.db.lock();
            automation_scheduler::complete_run(&db, &run.id, "completed", Some(&result_data))?;

            // Auto-archive if no actionable output
            if !has_actionable_output {
                let archive_updates = serde_json::json!({ "status": "archived" });
                let _ = auto_queries::update_automation_run(&db, &run.id, &archive_updates);
            }
        }

        // 6. Notify user
        send_notification(
            &app_handle,
            state,
            &automation.name,
            "success",
            &format!("Automation \"{}\" completed", automation.name),
        );

        info!(
            target: "automation-executor",
            "Run {} completed for {}", run.id, automation.name
        );

        Ok(())
    }
    .await;

    if let Err(e) = result {
        let message = e.to_string();

        {
            let db = state.db.lock();
            let error_result = serde_json::json!({ "error": message });
            let _ = automation_scheduler::complete_run(&db, &run.id, "failed", Some(&error_result));
        }

        send_notification(
            &app_handle,
            state,
            &automation.name,
            "error",
            &format!("Automation \"{}\" failed: {}", automation.name, message),
        );

        error!(
            target: "automation-executor",
            "Run {} failed for {}: {}", run.id, automation.name, message
        );

        return Err(AppError::Agent(message));
    }

    Ok(())
}

/// Send a notification to the frontend via Tauri events.
fn send_notification(
    app_handle: &AppHandle,
    state: &AppState,
    title: &str,
    notification_type: &str,
    message: &str,
) {
    let notification = AppNotification {
        id: uuid::Uuid::new_v4().to_string(),
        notification_type: notification_type.to_string(),
        title: title.to_string(),
        message: message.to_string(),
        timestamp: chrono::Utc::now().to_rfc3339(),
    };

    // Store in notification history
    {
        let mut notifications = state.notifications.lock();
        notifications.insert(0, notification.clone());
        if notifications.len() > 100 {
            notifications.truncate(100);
        }
    }

    let _ = app_handle.emit("notification", &notification);
}
